src/engine.ts

```diff
@@ -0,0 +1,207 @@
+    };
+  }
+}
+
+function partitionGroup(text: string, group: Group): Node[] {
+  const partitioned = group.others.reduce(({ inner: innerAcc, outer: outerAcc }, node) => {
+    const slicedNode = sliceNode(text, group.elected, node);
+    return {
+      inner: innerAcc.concat(slicedNode.inner),
+      outer: slicedNode.outer ? outerAcc.concat(slicedNode.outer) : outerAcc,
+    };
+  }, { inner: new Array<Node>(), outer: new Array<Node>()});
+
+  const { inner, outer } = partitioned;
+  const head = R.assoc('children', buildTree(text, inner), group.elected);
+  return [head].concat(buildTree(text, outer));
+}
+
+function groupNodes(nodes: Node[]): Node[][] {
+  const sortByStart = (nodeA: Node, nodeB: Node) => nodeA.start < nodeB.start;
+  const sortByEnd = (nodeA: Node, nodeB: Node) => nodeA.end < nodeB.end;
+  const sortedNodes = R.sortWith([sortByStart, sortByEnd], nodes);
+  return R.groupWith((nodeA, nodeB) => nodeA.end > nodeB.start, sortedNodes);
+}
+
+function electNode(candidates: Node[]): Group {
+  if (candidates.length === 0) {
+    throw new Error('Unable to elect node on empty list');    
+  } else {
+    const [elected, ...others] = R.sortBy(node => PRIORITIES[node.tag], candidates);
+    return { elected, others };
+  }
+}
+
+function fill(text: string, nodes: Node[], lowerBound: number, upperBound: number): Node[] {
+  return nodes.reduce((acc: Node[], node, index) => {
+    if (index === 0 && node.start > lowerBound) {
+      const textNode = new TextNode(lowerBound, node.start, text.slice(lowerBound, node.start));
+      return [textNode, node];
+    } else if (index === nodes.length - 1 && upperBound > node.end) {
+      const textNode = new TextNode(node.end, upperBound, text.slice(node.end, upperBound));
+      return acc.concat([node, textNode]);
+    } else {
+      const previousNode = nodes[index - 1];
+      if (previousNode) {
+        const textStart = previousNode.end;
+        const textEnd = node.start;
+
+        if (textEnd > textStart) {
+          const subtext = text.slice(textStart, textEnd);
+          const textNode = new TextNode(textStart, textEnd, subtext);
+          return acc.concat([textNode, node]);
+        }
+      }
+      return acc.concat(node);
+    }
+  }, []);
+}
+
+function buildTree(text: string, nodes: Node[], maybeLowerBound?: number, maybeUpperBound?: number): Node[] {
+  if (nodes.length > 0) {
+    const sortedNodes = R.sortBy(node => node.start, nodes);
+    const groups = groupNodes(sortedNodes);
+    const postElection = groups.map(electNode);
+    const tree = R.flatten(postElection.map(group => partitionGroup(text, group)));
+    const sortedTree = R.sortBy(node => node.start, tree);
+    const lowerBound = maybeLowerBound === undefined ? nodes[0].start : maybeLowerBound;
+    const upperBound = maybeUpperBound === undefined ? nodes[nodes.length - 1].end : maybeUpperBound;
+    return fill(text, sortedTree, lowerBound, upperBound);
+  } else {
+    return [];
```

Comment:
You probably miss something here. You should return an array with the text node.

Reply:
yep